Modules/_asynciomodule.c

```diff
@@ -119,26 +119,26 @@ typedef enum {
     PyObject *prefix##_result;                                              \
     PyObject *prefix##_source_tb;                                           \
     PyObject *prefix##_cancel_msg;                                          \
-    fut_state prefix##_state;                                               \
-    int prefix##_log_tb;                                                    \
-    int prefix##_blocking;                                                  \
     PyObject *dict;                                                         \
     PyObject *prefix##_weakreflist;                                         \
-    PyObject *prefix##_cancelled_exc;
+    PyObject *prefix##_cancelled_exc;                                       \
+    fut_state prefix##_state;                                               \
+    unsigned prefix##_log_tb: 1;                                            \
+    unsigned prefix##_blocking: 1;
 
 typedef struct {
     FutureObj_HEAD(fut)
 } FutureObj;
 
 typedef struct {
     FutureObj_HEAD(task)
+    unsigned task_must_cancel: 1;
+    unsigned task_log_destroy_pending: 1;
+    int task_num_cancels_requested;
```

Comment:
is my understand here correct, that this makes the alignment of `TaskObj` depend on the bitfield definition that's spread across the end of `FutureObj_HEAD` and the fields that immediately follow it here?

this feels fragile. not sure if there's a less fragile way to do this, but it at least warrants a comment documenting this coupling (in both places).

Reply:
I'm not sure what you feel is fragile about it. Surely this object definitions isn't part of any kind of ABI (or even API)?

Or are you worried that the presence of extra bitfields could change the alignment of *previous* bitfields? I.e. in
```
struct A {
    // Some fields
    unsigned a: 1;
}
struct AB {
    // Some fields, identical to above
    unsigned a: 1;
    unsigned b: 1;
}
```
do you think the exact placement for bit `a` could be different in `A` than in `AB`?

I believe that _in general_ the C standard requires a struct that extends another struct to use the same alignment for the common fields, but I don't know if there's an exception for bitfields...

Reply:
no ABI/API concerns

more about how the alignment of the bitfields of the task struct relies on the bitfields always being last in the future HEAD and immediately followed by the bitfields in the task struct

also seems more susceptible to tricky compiler bugs (or edge cases, or undefined behaviors) when accessing the last two fields from the HEAD definition on a future struct vs on a task struct (or worse, treating a pointer as either a fut or a struct and relying on the position of these fields to be consistent between the two, which may be true and may not be true, depending on compiler decisions and endianness etc)

Reply:
Yeah, the concerns would be:
- We could accidentally lose the space optimization if the header macro and the task field definitions changed in ways that caused a gap to appear in the alignment, and this would not be apparent to the maintainer adding the new field. This can be mitigated by adding comments in both places.
- The compiler (through a bug or by design) might place the bitfield in a different place in future vs. task. This would be disastrous, since the whole point of having the header macro in the first place is being able to cast a task pointer to a future pointer (and possibly back, if the type warrants it). We may be able to mitigate this through tests.

Reply:
It is only ever casted from task to future not the other way around.

Reply:
I am not a language lawyer nor compiler expert, but I have poked some folks who are (compiler experts, not language lawyers :) ), and they generally agree that this is probably in "undefined behavior" territory.

there are some guarantees about struct members layout when unions are involved (AND the union is visible in the scope of use) (ref: C spec section 6.5.2.3). there's a specific *invalid* example given in the spec (same section) that raises concerns about the ability to cast a task pointer to a future pointer regardless of the bitfield (it doesn't apply directly because it involves unions, but still concerning).

the "safe" recommendation was to use the future struct as the first member of the task struct:

```
typedef struct {
  // ... common members
} FutureObj;

typedef struct
  FutureObj fut;
  // ... task members
} TaskObj;
```

I think this will lose some of the memory savings (by not letting the bitfields from the end of the future and the beginning of the task to share a word), but will gain safety and some extensibility and readability (e.g. can move the task bitfield back to the end, and add more bitfields to either of them independently).

according to the spec, it would be safe to cast a task pointer to a future pointer (from https://en.cppreference.com/w/c/language/struct)

> Within a struct object, addresses of its elements (and the addresses of the bit-field allocation units) increase in order in which the members were defined. A pointer to a struct can be cast to a pointer to its first member (or, if the member is a bit-field, to its allocation unit). Likewise, a pointer to the first member of a struct can be cast to a pointer to the enclosing struct. There may be unnamed padding between any two members of a struct or after the last member, but not before the first member. 

beyond losing some of the memory reduction, it would also make it less convenient to access fut members on a task instance (require going through the `fut` member) (maybe that can be worked around with a union?)

in any case, this is the conclusion of my digging. I wouldn't object to leaving things as they are (with added comments) if we can also add tests that would catch issues with inconsistent struct memory layouts.